### PR TITLE
"param ** value" optimization for params that are positive powers of 2

### PR DESCRIPTION
**Pull request by Ldelaney**
Adds a version of the ** operator for the case where the LHS (base) is known at compile time and a positive power of two. 

Uses bit-shifting to efficiently raise the power. Uses complement and compare as an efficient method of testing whether the param is a power of 2.

Adds a performance test (possibly unneeded?) that compares the ** and << operators. Adds a correctness test to check that for bases 1-4 and powers 1-15, all of the results are correct. 

This may change the value returned if an overflow occurs, as the << operator will wrap around. 

I have a couple of comments about how things work so that people who don’t understand bits well can follow along. Should internal module code have comments beyond general section comments?

### CODE REVIEW

**Comment by Ldelaney** [comment]
@psahabu 
I probably only need the correctness test. I used the performance tests to confirm that there was a performance impact. The only value in keeping them would be to track for a performance regression in the future, but I find that unlikely, and think that adding a graph for this specific test may be overkill.

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -363,11 +363,27 @@ module ChapelBase {
       compilerError("unexpected case in exponentiation optimization");
   }
 
+  inline proc _expBaseHelp(param a: integral, b: int(?w)){
```
Comment (psahabu) [inline modules/internal/ChapelBase.chpl]: As far as I remember, you don't need the ?w if you're not using w. (I could be wrong.)

Comment (Ldelaney) [inline modules/internal/ChapelBase.chpl]: I think you are right here. _expHelp doesn't use the ?w either. I'll fix that.

**Comment by psahabu** [comment]
Yeah you can axe the perf test altogether. You can keep param-exp-shift.perfkeys though.

I'm not sure about my ?w comment, my feeling is that you can get rid of it but if it keeps you in line with the rest of the style in ChapelBase I wouldn't worry about it too much.

Otherwise looks good.

**Comment by Ldelaney** [comment]
Why keep the param-exp-shift.perfkeys?

**Comment by psahabu** [comment]
Hm, I have a vague recollection that having a perf mode on correctness tests is encouraged, but that may be a whole-cloth fabrication. Either way is fine, I just figured it's already there so it doesn't hurt much to keep it.

**Comment by Ldelaney** [comment]
param-exp-perf and param-exp-shift are both performance tests, and were used to compare the performance of ** vs << before and after the optimization. param-exp-correct is the only true correctness test in the directory. (It tests to ensure that 2^x, 3^x, and 4^x all work, which checks 2, a number that is not a power of 2, and a number other than 2 that is a power of 2)

At this point param-exp-shift isn't doing much.

**Comment by psahabu** [comment]
Ah sorry, I was mixing up the tests. Go ahead and delete that perfkey then.

**Pull request merged**